models/src/main/scala/coop/rchain/models/SortedParHashSet.scala

```diff
@@ -1,41 +1,46 @@
 package coop.rchain.models
 
-import scala.collection.generic.CanBuildFrom
-import scala.collection.immutable.HashSet
-import scala.collection.{mutable, SetLike}
+import coop.rchain.models.rholang.sorter.Sortable
 import coop.rchain.models.rholang.sorter.ordering._
+import monix.eval.Coeval
+
+import scala.collection.GenSet
+import scala.collection.immutable.HashSet
 
 //Enforce ordering and uniqueness.
 // - uniqueness is handled by using HashSet.
 // - ordering comes from sorting the elements prior to serializing.
-case class SortedParHashSet(private val ps: HashSet[Par])
-    extends Set[Par]
-    with SetLike[Par, SortedParHashSet] {
+final class SortedParHashSet(ps: HashSet[Par]) extends Iterable[Par] {
 
-  lazy val sortedPars: List[Par] = ps.toList.sort
+  lazy val sortedPars: List[Par]          = ps.toList.sort
+  private lazy val sortedPs: HashSet[Par] = HashSet(sortedPars: _*)
 
-  override def equals(that: Any): Boolean = that match {
-    case shs: SortedParHashSet => shs.sortedPars == this.sortedPars
-    case _                     => false
-  }
+  def +(elem: Par): SortedParHashSet = SortedParHashSet(ps + sort(elem))
 
-  override def hashCode(): Int = sortedPars.hashCode()
+  def -(elem: Par): SortedParHashSet = SortedParHashSet(ps - sort(elem))
 
-  override def empty: SortedParHashSet = SortedParHashSet.empty
+  def contains(elem: Par): Boolean = sortedPs.contains(sort(elem))
 
-  override def contains(elem: Par): Boolean = ps.contains(elem)
+  def union(that: GenSet[Par]): SortedParHashSet = SortedParHashSet(sortedPs.union(that.map(sort)))
 
-  override def +(elem: Par): SortedParHashSet = SortedParHashSet(ps + elem)
+  def empty: SortedParHashSet = SortedParHashSet(HashSet.empty[Par])
 
-  override def -(elem: Par): SortedParHashSet = SortedParHashSet(ps - elem)
+  def iterator: Iterator[Par] = sortedPars.toIterator
 
-  override def iterator: Iterator[Par] = sortedPars.toIterator
+  override def equals(that: Any): Boolean = that match {
+    case sph: SortedParHashSet => sph.sortedPars == this.sortedPars
+    case _                     => false
+  }
 
+  override def hashCode(): Int = sortedPars.hashCode()
+
+  private def sort(par: Par): Par = Sortable[Par].sortMatch[Coeval](par).map(_.term).value()
 }
 
 object SortedParHashSet {
-  def apply(seq: Seq[Par]): SortedParHashSet = SortedParHashSet(HashSet[Par](seq: _*))
-  def apply(): SortedParHashSet              = SortedParHashSet(HashSet.empty[Par])
+  def apply(seq: Seq[Par]): SortedParHashSet = new SortedParHashSet(HashSet[Par](seq: _*))
+
+  def apply(set: Set[Par]): SortedParHashSet = new SortedParHashSet(HashSet[Par](set.toSeq: _*))
```

Comment:
That's a total nit and feel free to omit that, but:

ideally, since we have the `apply` method variants on the companion, we'd only call the `new SortedParHashSet(...)` constructor once in the whole codebase. Each of the apply methods would either delegate to some other apply method variant, or would otherwise be the single method that uses `new`. I'm bringing this up only because I've seen bugs lurking in different possible initialization paths in my career, and the pattern I'm suggesting prevents that by delegating everything that's possible to the other factory methods.